osu.Game/Online/UserStatisticsWatcher.cs

```diff
@@ -98,31 +72,20 @@ private void userScoreProcessed(int userId, long scoreId)
             if (!watchedScores.Remove(scoreId, out var scoreInfo))
                 return;
 
-            requestStatisticsUpdate(userId, scoreInfo);
-        }
-
-        private void requestStatisticsUpdate(int userId, ScoreInfo scoreInfo)
-        {
-            var request = new GetUserRequest(userId, scoreInfo.Ruleset);
-            request.Success += user => Schedule(() => dispatchStatisticsUpdate(scoreInfo, user.Statistics));
-            api.Queue(request);
+            scorePendingUpdate = scoreInfo;
+            statisticsProvider.RefetchStatistics(scoreInfo.Ruleset);
         }
 
-        private void dispatchStatisticsUpdate(ScoreInfo scoreInfo, UserStatistics updatedStatistics)
+        private void onStatisticsUpdated(UserStatisticsUpdate update) => Schedule(() =>
         {
-            string rulesetName = scoreInfo.Ruleset.ShortName;
-
-            api.UpdateStatistics(updatedStatistics);
-
-            if (latestStatistics == null)
+            if (scorePendingUpdate == null || !update.Ruleset.Equals(scorePendingUpdate.Ruleset))
                 return;
 
-            latestStatistics.TryGetValue(rulesetName, out UserStatistics? latestRulesetStatistics);
-            latestRulesetStatistics ??= new UserStatistics();
+            if (update.OldStatistics != null)
+                latestUpdate.Value = new ScoreBasedUserStatisticsUpdate(scorePendingUpdate, update.OldStatistics, update.NewStatistics);
```

Comment:
I find this quite dodgy. What's the guarantee that you get the correct statistics update here and not some unrelated update queued by someone else later? Or that some unrelated statistics update isn't *treated* as a "score based" stats update?

If you ask me, `LocalUserStatisticsProvider.RefetchStatistics()` should return a `Task<UserStatisticsUpdate>`, so that task can be stored, and this component can be _sure_ that it's handling the data it explicitly requested.

Reply:
Since we want to be aware of which update is tied to the requested statistics refetch, wouldn't it be more easy to let the function accept a callback and invoke it directly from within the `Success` event of the request? I understand this heavily involves delegate variable capturing, but I think it's still within reason since requests are transient and die immediately once they're fulfilled, so there shouldn't be any room for potential event leaks? I'm not sure, I might be severely misunderstanding this. Here's an example diff of my proposal:
```diff
diff --git a/osu.Game/Online/LocalUserStatisticsProvider.cs b/osu.Game/Online/LocalUserStatisticsProvider.cs
index a25f5b05aa..1d322614e1 100644
--- a/osu.Game/Online/LocalUserStatisticsProvider.cs
+++ b/osu.Game/Online/LocalUserStatisticsProvider.cs
@@ -3,7 +3,9 @@
 
 using System;
 using System.Collections.Generic;
+using System.Diagnostics;
 using System.Linq;
+using System.Threading.Tasks;
 using osu.Framework.Allocation;
 using osu.Framework.Graphics;
 using osu.Game.Extensions;
@@ -36,7 +38,6 @@ public partial class LocalUserStatisticsProvider : Component
         private IAPIProvider api { get; set; } = null!;
 
         private readonly Dictionary<string, UserStatistics> statisticsCache = new Dictionary<string, UserStatistics>();
-        private readonly Dictionary<string, GetUserRequest> statisticsRequests = new Dictionary<string, GetUserRequest>();
 
         /// <summary>
         /// Returns the <see cref="UserStatistics"/> currently available for the given ruleset.
@@ -62,23 +63,21 @@ private void initialiseStatistics()
                 RefetchStatistics(ruleset);
         }
 
-        public void RefetchStatistics(RulesetInfo ruleset)
+        public void RefetchStatistics(RulesetInfo ruleset, Action<UserStatisticsUpdate>? callback = null)
         {
-            if (statisticsRequests.TryGetValue(ruleset.ShortName, out var previousRequest))
-                previousRequest.Cancel();
-
-            var request = statisticsRequests[ruleset.ShortName] = new GetUserRequest(api.LocalUser.Value.Id, ruleset);
-            request.Success += u => UpdateStatistics(u.Statistics, ruleset);
+            var request = new GetUserRequest(api.LocalUser.Value.Id, ruleset);
+            request.Success += u => UpdateStatistics(u.Statistics, ruleset, callback);
             api.Queue(request);
         }
 
-        protected void UpdateStatistics(UserStatistics newStatistics, RulesetInfo ruleset)
+        protected void UpdateStatistics(UserStatistics newStatistics, RulesetInfo ruleset, Action<UserStatisticsUpdate>? callback = null)
         {
             var oldStatistics = statisticsCache.GetValueOrDefault(ruleset.ShortName);
-
-            statisticsRequests.Remove(ruleset.ShortName);
             statisticsCache[ruleset.ShortName] = newStatistics;
-            StatisticsUpdated?.Invoke(new UserStatisticsUpdate(ruleset, oldStatistics, newStatistics));
+
+            var update = new UserStatisticsUpdate(ruleset, oldStatistics, newStatistics);
+            callback?.Invoke(update);
+            StatisticsUpdated?.Invoke(update);
         }
     }
 
diff --git a/osu.Game/Online/UserStatisticsWatcher.cs b/osu.Game/Online/UserStatisticsWatcher.cs
index 8ed1ff594d..e9b7c8e652 100644
--- a/osu.Game/Online/UserStatisticsWatcher.cs
+++ b/osu.Game/Online/UserStatisticsWatcher.cs
@@ -2,8 +2,11 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading.Tasks;
 using osu.Framework.Allocation;
 using osu.Framework.Bindables;
+using osu.Framework.Extensions;
 using osu.Framework.Extensions.ObjectExtensions;
 using osu.Framework.Graphics;
 using osu.Game.Extensions;
@@ -23,8 +26,6 @@ public partial class UserStatisticsWatcher : Component
         public IBindable<ScoreBasedUserStatisticsUpdate?> LatestUpdate => latestUpdate;
         private readonly Bindable<ScoreBasedUserStatisticsUpdate?> latestUpdate = new Bindable<ScoreBasedUserStatisticsUpdate?>();
 
-        private ScoreInfo? scorePendingUpdate;
-
         [Resolved]
         private SpectatorClient spectatorClient { get; set; } = null!;
 
@@ -43,7 +44,6 @@ protected override void LoadComplete()
             base.LoadComplete();
 
             spectatorClient.OnUserScoreProcessed += userScoreProcessed;
-            statisticsProvider.StatisticsUpdated += onStatisticsUpdated;
         }
 
         /// <summary>
@@ -72,21 +72,13 @@ private void userScoreProcessed(int userId, long scoreId)
             if (!watchedScores.Remove(scoreId, out var scoreInfo))
                 return;
 
-            scorePendingUpdate = scoreInfo;
-            statisticsProvider.RefetchStatistics(scoreInfo.Ruleset);
+            statisticsProvider.RefetchStatistics(scoreInfo.Ruleset, u => Schedule(() =>
+            {
+                if (u.OldStatistics != null)
+                    latestUpdate.Value = new ScoreBasedUserStatisticsUpdate(scoreInfo, u.OldStatistics, u.NewStatistics);
+            }));
         }
 
-        private void onStatisticsUpdated(UserStatisticsUpdate update) => Schedule(() =>
-        {
-            if (scorePendingUpdate == null || !update.Ruleset.Equals(scorePendingUpdate.Ruleset))
-                return;
-
-            if (update.OldStatistics != null)
-                latestUpdate.Value = new ScoreBasedUserStatisticsUpdate(scorePendingUpdate, update.OldStatistics, update.NewStatistics);
-
-            scorePendingUpdate = null;
-        });
-
         protected override void Dispose(bool isDisposing)
         {
             if (spectatorClient.IsNotNull())
```

I didn't go for it at first because I didn't think I need to care about which update is which in `UserStatisticsWatcher`, but here we are.

Reply:
I suppose a callback is the same thing. I'm not sure I have a strong preference of one over the other.